JOSS paper (#387)
SpeedyWeather/SpeedyWeather.jl: docs/joss/paper.md

---
title: 'SpeedyWeather.jl: Reinventing atmospheric general circulation models towards interactivity, extensibility and composability'

tags:
  - Julia
  - weather
  - climate
  - general circulation model
  - spectral
  - spherical harmonic transform

authors:
  - name: Milan Klöwer
    orcid: 0000-0002-3920-4356
    email: milank@mit.edu
    corresponding: true # (This is how to denote the corresponding author)
    affiliation: "1, 2" # (Multiple affiliations must be quoted)

  - name: Maximilian Gelbrecht
    affiliation: "3,4"

>   - name: Daisuke Hotta
>     affiliation: "5,6"

Reviewer: ```suggestion
  - name: Daisuke Hotta
    orcid: 0000-0003-2287-0608
    affiliation: "5,7"
```

  - name: Justin Willmert

>     affiliation: 7

Reviewer: ```suggestion
    affiliation: 6
```

  - name: Simone Silvestri
    affiliation: 1

>   - name: Gregory L Wagner

Reviewer 2: ```suggestion
  - name: Gregory L Wagner
    orcid: 0000-0001-5317-2445
```

    affiliation: 1

  - name: Alistair White
    affiliation: "3,4"

  - name: Sam Hatfield

>     affiliation: 8

Reviewer: ```suggestion
    affiliation: 7
```

  - name: David Meyer

>     affiliation: 8

Reviewer: ```suggestion
    affiliation: 7
```

  - name: Tom Kimpson

>     affiliation: "2,9"

Reviewer: ```suggestion
    affiliation: "2,8"
```

>   - name: Navid C Constantinou

Reviewer 2: ```suggestion
  - name: Navid C Constantinou
    orcid: 0000-0002-8149-4094
```

>     affiliation: 10

Reviewer: ```suggestion
    affiliation: 9
```

  - name: Chris Hill
    affiliation: 1
    
affiliations:
 - name: Massachusetts Institute of Technology, Cambridge, MA, USA
   index: 1
 - name: University of Oxford, UK
   index: 2
 - name: Technical University Munich, Germany
   index: 3
 - name: Potsdam Institute for Climate Impact Research, Germany
   index: 4
 - name: Japan Meteorological Agency, Tsukuba, Japan
   index: 5

>  - name: University of Reading, UK
>    index: 6

Reviewer: ```suggestion

```

 - name: University of Minnesota, Minneapolis, MN, USA

>    index: 7

Reviewer: ```suggestion
   index: 7
```
```suggestion
   index: 6
```

Author: No, you're 5,6 justin is 7 ;)

 - name: European Centre for Medium-Range Weather Forecasts, Reading, UK

>    index: 8

Reviewer: ```suggestion
   index: 7
```

 - name: University of Melbourne, Australia

>    index: 9

Reviewer: ```suggestion
   index: 8
```

 - name: Australian National University, Canberra, Australia

>    index: 10

Reviewer: ```suggestion
   index: 10
```
```suggestion
   index: 9
```

the shallow water equations (\autoref{fig:swm}), or the barotropic vorticity equations
with spherical harmonics. Several simple parameterizations for unresolved physical processes
such as precipitation or the boundary layer are implemented, and new ones can
be externally defined and passed as an argument to the model constructor.
SpeedyWeather.jl is an intermediate-complexity general circulation model [@Kucharski2013]
and research playground with an (almost) everything-flexible attitude.
It can be thought of as a conceptual reinvention of the Fortran SPEEDY model [@Molteni2003]
in the Julia programming language [@Bezanson2017].

SpeedyWeather.jl internally uses three sub-modules `SpeedyTransforms`, `RingGrids`, and
`LowerTriangularMatrices` to perform spherical harmonic transforms and interpolations
between various grids and the spectral space. `RingGrids` discretize the sphere
on iso-latitude rings and the spectral space is defined by the `LowerTriangularMatrices`
of the spherical harmonic coefficients. These three modules are independently usable
and therefore make SpeedyWeather.jl, beyond its main purpose of simulating the weather,

> also a library for the analysis of gridded data on the sphere.

Reviewer 2: If these three submodules are independently usable would it makes sense to have them as separate repositories in the SpeedyWeather organization?

Author: Yes, that is eventually the plan. However, I felt like this should wait a bit to make sure how these modules interact is a bit more established. Like I wanted to avoid that if we'd need to patch something in RingGrids, we have to wait for CI, release before having it also avaible in SpeedyWeather

Reviewer 2: I see. Yeah perhaps single GPU is a good first goal. But also no need to mention anything in the paper then.

> Running and analysing simulations can interactively combined, enhancing user

Reviewer: ```suggestion
Running and analysing simulations can be interactively combined, enhancing user
```

experience and productivity.

The user interface of SpeedyWeather.jl is heavily influenced by
the Julia ocean model Oceananigans.jl [@Ramadhan2020].

> A monolithic interface is deliberately avoided,

Reviewer 2: I don't understand what "monolithic" means in this context. Perhaps I'm not the only one? I suggest a small explanatory phrase next to monolithic the first time it appears.

Reviewer 3: "monolithic interface based on parameter files is avoided in favor of... [an library-style interface in which users write code to run models rather than merely supplying parameters and input arrays]"

Author: Good point to explain monolithic a bit more. Thanks Greg, I've included this sentence.

instead, a model is created bottom-up by first defining the discretization
and any non-default model components with its respective parameters.
All components are then collected into a single model object, which, once
initialized, returns a simulation object that contains the entire model state,
work arrays and parameters, that can be run, analysed or changed.
While these steps can be written into a script for reproducibility,
the same steps can be executed and interacted with one-by-one in
Julia's read-evaluate-print loop (REPL). We thereby reach an interactivity
far beyond a monolithic interface that is limited to the options provided.

> To be extendible and composable with new

Reviewer: ```suggestion
To be extensible and composable with new
```

One then only needs to extend the `initialize!` and `condensation!`
functions for this new type. Passing on `condensation = MyCondensation()`
to the model constructor then implements this new model component without
the need to branch off or overwrite existing model components.
Conceptually similar scientific modelling paradigms have been very successful
in the Python-based generic partial differential equation solver Dedalus [@Burns2020]
and the Julia ocean model Oceananigans.jl [@Ramadhan2020].

![Surface temperature simulated with the primitive equation model in SpeedyWeather.jl.
(Figure will be updated) \label{fig:primitive}](primitive.jpg)

The dynamical core of SpeedyWeather.jl uses established numerics
[@Bourke1972; @Hoskins1975; @Simmons1978; @Simmons1981],
widely adapted in numerical weather prediction. It is based on the spherical
harmonic transform with a leapfrog-based semi-implicit time integration [@Hoskins1975]
and a Robert-Asselin-Williams filter [@Williams2011; @Amezcua2011].
The spherical harmonic transform is grid-flexible. Any iso-latitude ring-based
grid can be used and new grids can be externally defined and passed on

> as argument. Many grids are already implemented: The conventional

Reviewer 2: ```suggestion
as argument. Many grids are already implemented: the conventional
```
?

Gaussian grid, a regular longitude-latitude grid, 
the octahedral Gaussian grid [@Malardel2016], the octahedral
Clenshaw-Curtis grid [@Hotta2018], and the HEALPix grid [@Gorski2005].
Both SpeedyWeather.jl and its spherical harmonic transform are also

> number format-flexible. 32-bit single-precision floating-point numbers

Reviewer 2: I don't understand the phrase "SpeedyWeather.jl and its spherical harmonic transform". Is SpeedyWeather a different entity of the transforms? Did we mean to write here that variables in both physical and spectral space are number format-flexible?

Author: It's technically two modules. You could achieve number format flexibility in SpeedyWeather by converting everything to Float32 before calling the transform if that was only working with Float32. Maybe adding `SpeedyTransforms` to highlight that I'm referring to the module?

(Float32) are the default as adapted by other modelling efforts [@Vana2017, @Nakano2018],
but Float64 and other custom number formats can be used with a single
code basis. Julia will compile to the choice of the number format, the grid,
and and other model components just-in-time. A simple parallelisation

> across vertical layers is supported with Julia's multi threading.

Reviewer 2: Hm... OK, so I read here that it's not parallelisable (eg MPI-type). Perhaps we can have a comment about "low memory allocation" (if that's true) "which allows for this and that..."

How about GPUs -- no mention about that? What's the state on running on GPUs (I am also asking out of curiosity, don't imply that we necessarily need to comment on the paper).

Author: Low memory allocation like Oceananigans can do it is really tricky with spectral models. At least I am not aware of a way to achieve that. As the transforms are costly and contain global communication you don't really have a bunch of stencils that can be fused to one...

Author: Some initial structure for GPUs is there, some Adapt calls, but no single kernel has actually been written for GPU. Maximilian experimented with a few in a branch but it's not in main. In general, I'm still not sure where SpeedyWeather should try to go. Multi-GPU or even Multi-CPU seems like a lot of work (domain decomposition in grid-point space and spectral space plus global communication), so I'm wondering whether a good goal for this model would be to aim for single GPU.

Reviewer 2: I see. Yeah perhaps single GPU is a good first goal. But also no need to mention anything  in the paper then given the current status of affairs.

Let alone online learning, which trains a neural network-based component together
with the rest of the model, accounting for interactions between components.
Gradients, necessary to optimize training, can be computed 
with automatic differentiation [@Moses2020], but only if differentiable functions
in a coherent language framework are provided.

We hope to provide with SpeedyWeather.jl a first test platform for data-driven
atmospheric modelling and in general an interactive model that makes difficult
problems easy to simulate. Climate models that are user-friendly, trainable,
but also easily extensible will suddenly make many complex
research ideas possible.

![Relative vorticity simulated with the shallow water model in SpeedyWeather.jl.
The simulation used a spectral resolution of T1023 (about 20km) and Float32
arithmetic on an octahedral Clenshaw-Curtis grid [@Hotta2018]. Relative vorticity
is visualised with Matplotlib [@Hunter2007] and Cartopy [@Cartopy] using a
transparent-to-white colormap to mimic the appearance of clouds. Underlain is
NASA's blue marble from June 2004. \label{fig:swm}](swm.png)

# Acknowledgements

> We acknowledge contributions from Mosè Giordano, Valentin Churavy and Pietro Monticone

Reviewer 2: ```suggestion
We acknowledge contributions from Mosè Giordano, Valentin Churavy, and Pietro Monticone
```

---
title: 'SpeedyWeather.jl: Reinventing atmospheric general circulation models towards interactivity, extensibility and composability'

tags:
  - Julia
  - weather
  - climate
  - general circulation model
  - spectral
  - spherical harmonic transform

authors:
  - name: Milan Klöwer
    orcid: 0000-0002-3920-4356
    email: milank@mit.edu
    corresponding: true # (This is how to denote the corresponding author)
    affiliation: "1, 2" # (Multiple affiliations must be quoted)

  - name: Maximilian Gelbrecht

>     affiliation: "3,4"

Reviewer 4: ```suggestion
    orchid: 0000-0002-0729-6671
    affiliation: "3,4"
```

Author: 🌷

  - name: Daisuke Hotta
    orcid: 0000-0003-2287-0608
    affiliation: "5,6"

  - name: Justin Willmert
    affiliation: 7

  - name: Simone Silvestri
    affiliation: 1

  - name: Gregory L Wagner
    orcid: 0000-0001-5317-2445
    affiliation: 1

  - name: Alistair White
    affiliation: "3,4"

  - name: Sam Hatfield
    affiliation: 6

  - name: David Meyer
    affiliation: 6

  - name: Tom Kimpson
    affiliation: "2,8"

  - name: Navid C Constantinou
    orcid: 0000-0002-8149-4094
    affiliation: 9

  - name: Chris Hill
    affiliation: 1
    
affiliations:
 - name: Massachusetts Institute of Technology, Cambridge, MA, USA
   index: 1
 - name: University of Oxford, UK
   index: 2

>  - name: Technical University Munich, Germany

Reviewer 4: ```suggestion
 - name: Technical University of Munich, Germany
```

and therefore make SpeedyWeather.jl, beyond its main purpose of simulating the weather,
also a library for the analysis of gridded data on the sphere.
Running and analysing simulations can be interactively combined, enhancing user
experience and productivity.

The user interface of SpeedyWeather.jl is heavily influenced by
the Julia ocean model Oceananigans.jl [@Ramadhan2020].
A monolithic interface based on parameter files is avoided in favor of a
library-style interface in which users write code to run models rather than
merely supplying parameters and input arrays.
A model is created bottom-up by first defining the discretization
and any non-default model components with its respective parameters.
All components are then collected into a single model object, which, once
initialized, returns a simulation object that contains the entire model state,
work arrays and parameters, that can be run, analysed or changed.
While these steps can be written into a script for reproducibility,
the same steps can be executed and interacted with one-by-one in
Julia's read-evaluate-print loop (REPL). We thereby reach an interactivity

> far beyond a monolithic interface that is limited to the options provided.

Reviewer 4: ```suggestion
far beyond a monolithic interface that is limited to the options provided. At the same time, sensible default arguments enable even inexperienced users, to run simulations in just a few lines of code. 
```

Besides the modularity / bottom-up approach, I'd very briefly also mention how easy it is to actually get the model running. I think this is worth mentioning because starting a run - even with default arguments - of some old Fortran models can be quite cumbersome.

Author: Agree!

To be extensible and composable with new
model components, SpeedyWeather.jl relies on Julia's multiple dispatch
programming paradigm [@Bezanson2017]. Every model component
is defined as a new type. For example, to define a new way how to calculate
the precipitation due to the physical process of large-scale condensation,
one would define `MyCondensation` as a new subtype of `AbstractCondensation`.
One then only needs to extend the `initialize!` and `condensation!`
functions for this new type. Passing on `condensation = MyCondensation()`
to the model constructor then implements this new model component without
the need to branch off or overwrite existing model components.
Conceptually similar scientific modelling paradigms have been very successful
in the Python-based generic partial differential equation solver Dedalus [@Burns2020]
and the Julia ocean model Oceananigans.jl [@Ramadhan2020].

![Surface temperature simulated with the primitive equation model in SpeedyWeather.jl.
(Figure will be updated) \label{fig:primitive}](primitive.jpg)

The dynamical core of SpeedyWeather.jl uses established numerics
[@Bourke1972; @Hoskins1975; @Simmons1978; @Simmons1981],

> widely adapted in numerical weather prediction. It is based on the spherical

Reviewer 5: ```suggestion
widely adopted in numerical weather prediction. It is based on the spherical
```

harmonic transform with a leapfrog-based semi-implicit time integration [@Hoskins1975]
and a Robert-Asselin-Williams filter [@Williams2011; @Amezcua2011].
The spherical harmonic transform is grid-flexible. Any iso-latitude ring-based

> grid can be used and new grids can be externally defined and passed on

Reviewer 5: ```suggestion
grid can be used and new grids can be externally defined and passed in
```

> as argument. Many grids are already implemented: the conventional

Reviewer 5: ```suggestion
as an argument. Many grids are already implemented: the conventional
```

Gaussian grid, a regular longitude-latitude grid, 
the octahedral Gaussian grid [@Malardel2016], the octahedral
Clenshaw-Curtis grid [@Hotta2018], and the HEALPix grid [@Gorski2005].
Both SpeedyWeather.jl and its spherical harmonic transform SpeedyTransforms are also
number format-flexible. 32-bit single-precision floating-point numbers

> (Float32) are the default as adapted by other modelling efforts [@Vana2017, @Nakano2018],

Reviewer 5: ```suggestion
(Float32) are the default as adopted by other modelling efforts [@Vana2017, @Nakano2018],
```

We hope to provide with SpeedyWeather.jl a first test platform for data-driven
atmospheric modelling and in general an interactive model that makes difficult
problems easy to simulate. Climate models that are user-friendly, trainable,
but also easily extensible will suddenly make many complex
research ideas possible.

![Relative vorticity simulated with the shallow water model in SpeedyWeather.jl.
The simulation used a spectral resolution of T1023 (about 20km) and Float32
arithmetic on an octahedral Clenshaw-Curtis grid [@Hotta2018]. Relative vorticity
is visualised with Matplotlib [@Hunter2007] and Cartopy [@Cartopy] using a
transparent-to-white colormap to mimic the appearance of clouds. Underlain is
NASA's blue marble from June 2004. \label{fig:swm}](swm.png)

# Acknowledgements

We acknowledge contributions from Mosè Giordano, Valentin Churavy, and Pietro Monticone
who have also committed to the SpeedyWeather.jl repository, and the wider Julia community
for help and support. We gratefully acknowledge funding from the 
National Science Foundation (Chris please add) and the European Research Council
under the European Union’s Horizon 2020 research and innovation programme

> for the ITHACA grant (no. 741112).

Reviewer 2: ```suggestion
for the ITHACA grant (no. 741112).
N.C.C. acknowledges support by the Australian Research Council DECRA Fellowship DE210100749.
```

Author: Added

tags:
  - Julia
  - weather
  - climate
  - general circulation model
  - spectral
  - spherical harmonic transform

authors:
  - name: Milan Klöwer
    orcid: 0000-0002-3920-4356
    email: milank@mit.edu
    corresponding: true # (This is how to denote the corresponding author)
    affiliation: "1, 2" # (Multiple affiliations must be quoted)

  - name: Maximilian Gelbrecht
    orcid: 0000-0002-0729-6671
    affiliation: "3,4"

  - name: Daisuke Hotta
    orcid: 0000-0003-2287-0608
    affiliation: "5,6"

  - name: Justin Willmert
    affiliation: 7

  - name: Simone Silvestri
    affiliation: 1

  - name: Gregory L Wagner
    orcid: 0000-0001-5317-2445
    affiliation: 1

  - name: Alistair White
    affiliation: "3,4"

  - name: Sam Hatfield
    affiliation: 6

  - name: David Meyer
    affiliation: 6

  - name: Tom Kimpson
    affiliation: "2,8"

  - name: Navid C Constantinou
    orcid: 0000-0002-8149-4094
    affiliation: 9

  - name: Chris Hill
    affiliation: 1
    
affiliations:
 - name: Massachusetts Institute of Technology, Cambridge, MA, USA
   index: 1

>  - name: University of Oxford, UK

Reviewer 6: Missing index:2 here for Oxford

Running and analysing simulations can be interactively combined, enhancing user
experience and productivity.

The user interface of SpeedyWeather.jl is heavily influenced by
the Julia ocean model Oceananigans.jl [@Ramadhan2020].
A monolithic interface based on parameter files is avoided in favor of a
library-style interface in which users write code to run models rather than
merely supplying parameters and input arrays.
A model is created bottom-up by first defining the discretization
and any non-default model components with its respective parameters.
All components are then collected into a single model object, which, once
initialized, returns a simulation object that contains the entire model state,
work arrays and parameters, that can be run, analysed or changed.
While these steps can be written into a script for reproducibility,
the same steps can be executed and interacted with one-by-one in
Julia's read-evaluate-print loop (REPL). We thereby reach an interactivity
far beyond a monolithic interface that is limited to the options provided.
At the same time, sensible default arguments enable even inexperienced users

> to run simulations in just a few lines of code. 

Author: Added that defaults are well-established test cases

To be extensible and composable with new
model components, SpeedyWeather.jl relies on Julia's multiple dispatch
programming paradigm [@Bezanson2017]. Every model component
is defined as a new type. For example, to define a new way how to calculate
the precipitation due to the physical process of large-scale condensation,
one would define `MyCondensation` as a new subtype of `AbstractCondensation`.
One then only needs to extend the `initialize!` and `condensation!`
functions for this new type. Passing on `condensation = MyCondensation()`
to the model constructor then implements this new model component without
the need to branch off or overwrite existing model components.
Conceptually similar scientific modelling paradigms have been very successful
in the Python-based generic partial differential equation solver Dedalus [@Burns2020]

> and the Julia ocean model Oceananigans.jl [@Ramadhan2020].

Author: Similar to my comments from above. Please justify why there is a need to reword the entire paragraph, especially because you introduced several points that are less correct, less concise as they were before. You write "distinct type", which I'd like to avoid because these types are sub and supertypes of each other, so they have a relation and depending on the type tree aren't always distinguished and can be subtyped. You write "functions need modification for this type" no, that's not how multiple dispatch works. There's a reason I wrote "extend" rather than "modify" because with multiple dispatch you don't have to modify or overwrite the previous definition of a function, rather you extend it by defining a new method, that's very different and in fact our entire approach wouldn't work if it wasn't for that!

![Surface temperature simulated with the primitive equation model in SpeedyWeather.jl.
(Figure will be updated) \label{fig:primitive}](primitive.jpg)

The dynamical core of SpeedyWeather.jl uses established numerics
[@Bourke1972; @Hoskins1975; @Simmons1978; @Simmons1981],
widely adopted in numerical weather prediction. It is based on the spherical
harmonic transform with a leapfrog-based semi-implicit time integration [@Hoskins1975]
and a Robert-Asselin-Williams filter [@Williams2011; @Amezcua2011].
The spherical harmonic transform is grid-flexible. Any iso-latitude ring-based
grid can be used and new grids can be externally defined and passed in
as an argument. Many grids are already implemented: the conventional
Gaussian grid, a regular longitude-latitude grid, 
the octahedral Gaussian grid [@Malardel2016], the octahedral

> Clenshaw-Curtis grid [@Hotta2018], and the HEALPix grid [@Gorski2005].

Author: Similar as before, I don't see the win over the new paragraph compared to the old, especially because we now have two very long sentences, whereas the previous version had a clear summary sentence to start with before going into the details.

Both SpeedyWeather.jl and its spherical harmonic transform SpeedyTransforms are also

> number format-flexible. 32-bit single-precision floating-point numbers

Author: No, no. SpeedyWeather could be number format flexible without its transform and vice versa. I've been playing with the idea to use one format for the dynamical core exclusive the transform, but always Float32 for the transform as FFTW only supports Float32/64

(Float32) are the default as adopted by other modelling efforts [@Vana2017; @Nakano2018],
but Float64 and other custom number formats can be used with a single
code basis [@Klower2022; @Klower2020].

> Julia will compile to the choice of the number format, the grid,

Author: Where do I say this is running in half precision? The FFT will switch between FFTW and GenericFFT depending on the format, using half precision is a precision issue not a compiler issue anymore. Other formats like posits may require certain methods that I haven't defined yet, but that's not a problem of SpeedyWeather.

and and other model components just-in-time. A simple parallelisation
across vertical layers is supported with Julia's multi threading.
Output is stored as NetCDF files using

> [NCDatasets.jl](https://github.com/Alexander-Barth/NCDatasets.jl).

Author: Most weather, atm, ocean and climate models continue to run in Float64 by default, those two are merely exceptions. So I'd leave the original formulation as is.

# Statement of need

SpeedyWeather.jl is a fresh approach to atmospheric
models that have been very influential in many areas of scientific 
and high-performance computing as well as climate change mitigation and adaptation.
Most weather, ocean and climate models are written in Fortran and have been developed over
decades. From this tradition follows a specific programming style and
associated user interface. Running a simulation in Fortran and analysing the
data in Python makes it virtually impossible to interact with various model
components interactively. Furthermore, data-driven climate modelling
[@Rasp2018; @Schneider2023], which replaces existing model components with machine learning
is difficult due to the lack of established deep learning frameworks in Fortran [@Innes2019].
Let alone online learning, which trains a neural network-based component together
with the rest of the model, accounting for interactions between components.
Gradients, necessary to optimize training, can be computed 
with automatic differentiation [@Moses2020], but only if differentiable functions

> in a coherent language framework are provided.

Author: I like the first sentences you wrote and will incorporate them in the next days. Can you clarify whether there's a difference between your paper and your thesis, or whether for the sake of this citation they actually refer to the same piece of work. Very happy to cite your paper!

> 

Author: How is "hope" an overstatement?

  - name: Chris Hill
    affiliation: 1
    
affiliations:
 - name: Massachusetts Institute of Technology, Cambridge, MA, USA
   index: 1
 - name: University of Oxford, UK
   index: 2
 - name: Technical University of Munich, Germany
   index: 3
 - name: Potsdam Institute for Climate Impact Research, Germany
   index: 4
 - name: Japan Meteorological Agency, Tsukuba, Japan
   index: 5
 - name: European Centre for Medium-Range Weather Forecasts, Reading, UK
   index: 6
 - name: University of Minnesota, Minneapolis, MN, USA
   index: 7
 - name: Imperial College London, UK
   index: 8
 - name: European Centre for Medium-Range Weather Forecasts, Bonn, Germany
   index: 9
 - name: University of Melbourne, Australia
   index: 10
 - name: Australian National University, Canberra, Australia
   index: 11
   
date: 14 September 2023
bibliography: paper.bib

---

# Summary

SpeedyWeather.jl is a library to simulate and analyze the global atmospheric
circulation on the sphere. It implements several 2D and 3D
models to solve the primitive equations with and without humidity (\autoref{fig:primitive}),
the shallow water equations (\autoref{fig:swm}), or the barotropic vorticity equations

> with spherical harmonics. Several simple parameterizations for unresolved physical processes

Reviewer 7: That said, I'm sure the meaning will be clear to a reader familiar with these models and methods.

between various grids and the spectral space. `RingGrids` discretize the sphere
on iso-latitude rings and the spectral space is defined by the `LowerTriangularMatrices`
of the spherical harmonic coefficients. These three modules are independently usable
and therefore make SpeedyWeather.jl, beyond its main purpose of simulating the weather,
also a library for the analysis of gridded data on the sphere.
Running and analysing simulations can be interactively combined, enhancing user
experience and productivity.

The user interface of SpeedyWeather.jl is heavily influenced by
the Julia ocean model Oceananigans.jl [@Ramadhan2020].
A monolithic interface based on parameter files is avoided in favor of a
library-style interface in which users write code to run models rather than
merely supplying parameters and input arrays.
A model is created bottom-up by first defining the discretization
and any non-default model components with their respective parameters.
All components are then collected into a single model object which, once
initialized, returns a simulation object that contains the entire model state,

> work arrays and parameters, that can be run, analysed or changed.

Reviewer 7: Also analyzed not analysed

(Figure will be updated) \label{fig:primitive}](primitive.jpg)

The dynamical core of SpeedyWeather.jl uses established numerics
[@Bourke1972; @Hoskins1975; @Simmons1978; @Simmons1981],
widely adopted in numerical weather prediction. It is based on the spherical
harmonic transform with a leapfrog-based semi-implicit time integration [@Hoskins1975]
and a Robert-Asselin-Williams filter [@Williams2011; @Amezcua2011].
The spherical harmonic transform is grid-flexible. Any iso-latitude ring-based
grid can be used and new grids can be externally defined and passed in
as an argument. Many grids are already implemented: the conventional
Gaussian grid, a regular longitude-latitude grid, 
the octahedral Gaussian grid [@Malardel2016], the octahedral
Clenshaw-Curtis grid [@Hotta2018], and the HEALPix grid [@Gorski2005].
Both SpeedyWeather.jl and its spherical harmonic transform SpeedyTransforms are also
number format-flexible. 32-bit single-precision floating-point numbers
(Float32) are the default as adopted by other modelling efforts [@Vana2017; @Nakano2018],
but Float64 and other custom number formats can be used with a single
code basis [@Klower2022; @Klower2020].

> Julia will compile to the choice of the number format, the grid,

Author: Added `the`

> Julia will compile to the choice of the number format, the grid,

Reviewer 7: I actually suggested removing `the` :sweat_smile:

> and and other model components just-in-time. A simple parallelisation

Reviewer 7: Still an extra `and`

> across vertical layers is supported with Julia's multi threading.

Author: changed

Output is stored as NetCDF files using
[NCDatasets.jl](https://github.com/Alexander-Barth/NCDatasets.jl).

# Statement of need

SpeedyWeather.jl is a fresh approach to atmospheric

> models that have been very influential in many areas of scientific 

Author: At least every other paper in the field uses a model directly or indirectly, does that not justify the "very"?

and high-performance computing as well as climate change mitigation and adaptation.
Most weather, ocean and climate models are written in Fortran and have been developed over
decades. From this tradition follows a specific programming style and
associated user interface. Running a simulation in Fortran and analysing the
data in Python makes it virtually impossible to interact with various model

> components interactively. Furthermore, data-driven climate modelling

Reviewer 7: I could be wrong but I think "modeling" is preferred over "modelling" in American English?

> [@Rasp2018; @Schneider2023], which replaces existing model components with machine learning

Reviewer 7: ```suggestion
[@Rasp2018; @Schneider2023], which replaces existing model components with machine learning,
```

is difficult due to the lack of established deep learning frameworks in Fortran [@Innes2019].

> Let alone online learning, which trains a neural network-based component together

Reviewer 7: I am fairly confident that a sentence shouldn't start with "let alone". Better if it's used as a conjunction in a subordinate clause, e.g. "X, let alone Y".

Author: Yes you're right. I just deleted that sentence I think it goes into unnecessary detail.

with the rest of the model, accounting for interactions between components.

> Gradients, necessary to optimize training, can be computed 

Reviewer 7: ```suggestion
Gradients, necessary to optimize neural network weights during training, can be computed 
```
Training is the process of optimization rather than the thing which is optimized.

with automatic differentiation [@Moses2020], but only if differentiable functions
in a coherent language framework are provided.

> We hope to provide with SpeedyWeather.jl a first test platform for data-driven

Reviewer 7: ```suggestion
With SpeedyWeather.jl we hope to provide a first test platform for data-driven
```
Just a suggestion, sounds better to my ear

Author: Sounds good!

atmospheric modelling and in general an interactive model that makes difficult
problems easy to simulate. Climate models that are user-friendly, trainable,
but also easily extensible will suddenly make many complex
research ideas possible.

![Relative vorticity simulated with the shallow water model in SpeedyWeather.jl.
The simulation used a spectral resolution of T1023 (about 20 km) and Float32
arithmetic on an octahedral Clenshaw-Curtis grid [@Hotta2018]. Relative vorticity

> is visualised with Matplotlib [@Hunter2007] and Cartopy [@Cartopy] using a

Reviewer 7: ```suggestion
is visualized with Matplotlib [@Hunter2007] and Cartopy [@Cartopy] using a
```

> transparent-to-white colormap to mimic the appearance of clouds. Underlain is

Reviewer 7: ```suggestion
transparent-to-white colormap to mimic the appearance of clouds. Underlaid is
```
I think you want the past particle of underlay rather than underlie, since you put the thing there.

Author: Haha, yes, English is difficult!

    orcid: 0000-0001-5317-2445
    affiliation: 1

  - name: Alistair White
  - orcid: 0000-0003-3377-6852
    affiliation: "3,4"

  - name: Sam Hatfield
    orcid: 0000-0001-7235-6450
    affiliation: 6

  - name: Tom Kimpson
    orcid: 0000-0002-6542-6032
    affiliation: "2,8"

  - name: Navid C Constantinou
    orcid: 0000-0002-8149-4094
    affiliation: 9

  - name: Chris Hill
    affiliation: 1
    
affiliations:
 - name: Massachusetts Institute of Technology, Cambridge, MA, USA
   index: 1
 - name: University of Oxford, UK
   index: 2
 - name: Technical University of Munich, Germany
   index: 3
 - name: Potsdam Institute for Climate Impact Research, Germany
   index: 4
 - name: Japan Meteorological Agency, Tsukuba, Japan
   index: 5
 - name: European Centre for Medium-Range Weather Forecasts, Reading, UK
   index: 6
 - name: University of Minnesota, Minneapolis, MN, USA
   index: 7
 - name: University of Melbourne, Australia
   index: 8
 - name: Australian National University, Canberra, Australia
   index: 9
   
date: 14 September 2023
bibliography: paper.bib

---

# Summary

SpeedyWeather.jl is a library to simulate and analyze the global atmospheric

> circulation on the sphere. It implements several 2D and 3D
> models solved with spherical harmonics:

Reviewer 8: ```suggestion
circulation on the sphere. It implements several 2D and 3D
spectral models solving different sets of equations:
```
spherical harmonics is only a 2D discretization method, not 3D, for this reason (and the fast that SpeedyWeather is more than spherical harmonics) I would not emphasize SPH here but when you talking about numerics, maybe where you introduce the modules in the second paragraph

- the primitive equations with and without humidity (\autoref{fig:primitive}),
- the shallow water equations (\autoref{fig:swm}), and
- the barotropic vorticity equations.

Several simple parameterizations for unresolved physical processes

> such as precipitation or the boundary layer are implemented, and new ones can

Reviewer 8: ```suggestion
such as precipitation or boundary layer mixing are implemented, and new ones can
```

be externally defined and passed as an argument to the model constructor.
SpeedyWeather.jl is an intermediate-complexity general circulation model [@Kucharski2013]
and research playground with an (almost) everything-flexible attitude.
It can be thought of as a conceptual reinvention of the Fortran SPEEDY model [@Molteni2003]
in the Julia programming language [@Bezanson2017].

> SpeedyWeather.jl internally uses three sub-modules `RingGrids`,

Reviewer 8: I would write here about SPH, something like this:
```suggestion
SpeedyWeather.jl approximated the governing equations on a sphere using a spectral approach in the horizontal direction (Spherical Harmonics) and a finite difference approach in the vertical direction. To achieve this goal, SpeedyWeather.jl makes use of three sub-modules `RingGrids`, ...
```
(feel free to change it as you like)

Author: After your restructure suggestions I feel this is no longer needed as this paragraph directly follows the paragraph on numerical methods.

`LowerTriangularMatrices`, and `SpeedyTransforms`. `RingGrids` is a module that discretizes
the sphere on iso-latitude rings and implements interpolations between various such grids.

> `LowerTriangularMatrices` is a module used to define the spectral space of the spherical

Reviewer 8: I think this is not super-clear. Maybe a clarifying sentence could be something like: 
"LowerTriangularMatrices facilitates the spherical harmonic implementation by organizing the SPH coefficients in a lower triangular matrix representation"
Disregard if I am wrong

harmonic coefficients. `SpeedyTransforms` implements the spectral transform between
the grid-point space as defined by `RingGrids` and the spectral space defined in
`LowerTriangularMatrices`. These three modules are independently usable
and therefore make SpeedyWeather.jl, beyond its main purpose of simulating 
atmospheric motion, also a library for the analysis of gridded data on the sphere.
Running and analyzing simulations can be interactively combined, enhancing user
experience and productivity.

The user interface of SpeedyWeather.jl is heavily influenced by
the Julia ocean model Oceananigans.jl [@Ramadhan2020].
A monolithic interface based on parameter files is avoided in favor of a
library-style interface. Users write notebooks, directly into
Julia's read-evaluate-print loop (REPL) or short scripts to run models
rather than merely supplying parameters and input arrays.

> A model is created bottom-up by first defining the discretization

Reviewer 8: "constructed" or "created"?

and any non-default model components with their respective parameters.
All components are then collected into a single model object which, once
initialized, returns a simulation object. A simulation contains everything,
the model with all parameters as created before but also all prognostic and diagnostic variables.
Such a simulation can then be run, but also accessed before and after to analyze or
visualize the current variables, or individual terms of the equations.

> One can also adjust parameters or define new model components before resuming the simulation.

Reviewer 8: I would interpret this as: "You can change the model as it is running". However, `Model`s (and `Simulation`s) are not mutable, so it is not actually possible to do this, no?

> While these steps can be written into a script for reproducibility,

Reviewer 8: Do you want to make the steps more similar to a list so here it is a bit clearer what you mean by steps? Like
"The classical SpeedyWeather workflow consists in: \textit{(i)} constructing a model bottom up by defining..., \textit{(ii)} initialize the model which will return a Simulation object containing the model numerics and parameterization with the desired parameters..., \textit{(iii)} run the model. During this step it is possible to pause..."

the same steps can be executed and interacted with one-by-one in
the REPL or in a single Jupyter or Pluto notebook.
We thereby achieve an interactivity of a simulation and its various model components
far beyond the options provided in a monolithic interface.
At the same time, defaults, set to well-established test cases, 
enable even inexperienced users to run simulations in just a few lines of code. 

> To be extensible with new model components, SpeedyWeather.jl relies on Julia's

Reviewer 8: Also here I think "new model components" is a bit ambiguous. I think you can add new parameterizations easily, with the condition that similar physics (aka the model component) is already implemented and accounted for right? I would clear a bit the air by specifying: "All model components can be extended to support the custom parameterization of choice by relying on Julia's multiple dispatch programming paradigm"

multiple dispatch programming paradigm [@Bezanson2017].

> Every model component is defined as a new type.

Reviewer 8: And here I would change this line to 
```suggestion
All supported model components (for example condensation, radiation, boundary-layer mixing) define an `Abstract` type that can be easily subtypes to introduce new parameterization. For example...
```

harmonic transform [@reinecke2013; @stompor2011] with a leapfrog-based semi-implicit
time integration [@Hoskins1975] and a Robert-Asselin-Williams filter [@Williams2011; @Amezcua2011].
The spherical harmonic transform is grid-flexible [@Willmert2020]. Any iso-latitude ring-based
grid can be used and new grids can be externally defined and passed in
as an argument. Many grids are already implemented: the conventional
Gaussian grid, a regular longitude-latitude grid, 
the octahedral Gaussian grid [@Malardel2016], the octahedral
Clenshaw-Curtis grid [@Hotta2018], and the HEALPix grid [@Gorski2005].
Both SpeedyWeather.jl and its spherical harmonic transform `SpeedyTransforms` are also
number format-flexible. Single-precision floating-point numbers
(Float32) are the default as adopted by other modelling efforts [@Vana2017; @Nakano2018],
but Float64 and other custom number formats can be used with a single
code basis [@Klower2022; @Klower2020].
Julia will compile to the choice of number format, the grid,
and and other model components just-in-time. A simple parallelization
across vertical layers is supported by Julia's multithreading.
Output is stored as NetCDF files using

> [NCDatasets.jl](https://github.com/Alexander-Barth/NCDatasets.jl).

Reviewer 8: I would probably connect this paragraph with the submodules paragraph as they treat similar topics. So either move the submodules paragraph before this one or this one before the submodules on (disregard if you prefer it like this)

> [NCDatasets.jl](https://github.com/Alexander-Barth/NCDatasets.jl).

Reviewer 8: ```suggestion
NCDatasets.jl (reference to NCDatasets.jl).
```

Reviewer 2: I don't think there is anything to cite for NCDatasets.jl @simone-silvestri, is there?

Reviewer 8: maybe something like this?
```
@misc{ncdatasets,
  publisher = {GitHub},
  journal = {GitHub repository},
  howpublished = {\url{https://github.com/Alexander-Barth/NCDatasets.jl}},
  commit = {4f57d6a0e4c030202a07a60bc1bb1ed1544bf679}
}
```
I am not sure if there is a standard way to cite software but I wouldn't hide it in a hyperlink.

Reviewer 2: OK, indeed! This is better!

Reviewer 2: x-ref https://github.com/Alexander-Barth/NCDatasets.jl/issues/234

Author: Added!

  - name: Daisuke Hotta
    orcid: 0000-0003-2287-0608
    affiliation: "5,6"

  - name: Justin Willmert
    orcid: 0000-0002-6452-4693
    affiliation: 7

  - name: Simone Silvestri
    affiliation: 1

  - name: Gregory L Wagner
    orcid: 0000-0001-5317-2445
    affiliation: 1

  - name: Alistair White
    orcid: 0000-0003-3377-6852
    affiliation: "3,4"

  - name: Sam Hatfield
    orcid: 0000-0001-7235-6450
    affiliation: 6

  - name: Tom Kimpson
    orcid: 0000-0002-6542-6032
    affiliation: "2,8"

  - name: Navid C Constantinou
    orcid: 0000-0002-8149-4094
    affiliation: 9

  - name: Chris Hill
    affiliation: 1
    
affiliations:
 - name: Massachusetts Institute of Technology, Cambridge, MA, USA
   index: 1
 - name: University of Oxford, UK
   index: 2
 - name: Technical University of Munich, Germany
   index: 3
 - name: Potsdam Institute for Climate Impact Research, Germany
   index: 4
 - name: Japan Meteorological Agency, Tsukuba, Japan
   index: 5
 - name: European Centre for Medium-Range Weather Forecasts, Reading, UK
   index: 6
 - name: University of Minnesota, Minneapolis, MN, USA
   index: 7
 - name: University of Melbourne, Australia
   index: 8

>  - name: Australian National University, Canberra, Australia

Reviewer 2: ```suggestion
 - name: The University of Melbourne, Parkville, VIC, Australia
```

Several simple parameterizations for unresolved physical processes
such as precipitation or the boundary layer are implemented, and new ones can
be externally defined and passed as an argument to the model constructor.
SpeedyWeather.jl is an intermediate-complexity general circulation model [@Kucharski2013]
and research playground with an (almost) everything-flexible attitude.
It can be thought of as a conceptual reinvention of the Fortran SPEEDY model [@Molteni2003]
in the Julia programming language [@Bezanson2017].

SpeedyWeather.jl internally uses three sub-modules `RingGrids`,
`LowerTriangularMatrices`, and `SpeedyTransforms`. `RingGrids` is a module that discretizes
the sphere on iso-latitude rings and implements interpolations between various such grids.
`LowerTriangularMatrices` is a module used to define the spectral space of the spherical
harmonic coefficients. `SpeedyTransforms` implements the spectral transform between
the grid-point space as defined by `RingGrids` and the spectral space defined in
`LowerTriangularMatrices`. These three modules are independently usable
and therefore make SpeedyWeather.jl, beyond its main purpose of simulating 

> atmospheric motion also a library for the analysis of gridded data on the sphere.

Reviewer 7: ```suggestion
atmospheric motion, also a library for the analysis of gridded data on the sphere.
```

> Running and analysing simulations can be interactively combined, enhancing user

Reviewer 7: ```suggestion
Running and analyzing simulations can be interactively combined, enhancing user
```
